os/kernel/binary_manager/binary_manager_resource.c

```diff
@@ -0,0 +1,174 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+#include <tinyara/config.h>
+#include <stdio.h>
+#include <debug.h>
+#include <stdlib.h>
+#include <queue.h>
+#include <errno.h>
+#include <mqueue.h>
+#ifdef CONFIG_FS_NAMED_SEMAPHORES
+#include <semaphore.h>
+#endif
+#include <sys/types.h>
+#include <tinyara/mm/mm.h>
+#include <tinyara/sched.h>
+#include <tinyara/init.h>
+
+#include "sched/sched.h"
+#include "binary_manager.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+struct resource_node_s {
+	struct resource_node_s *flink;
+	int type;
+	char name[CONFIG_NAME_MAX];
+};
+typedef struct resource_node_s resource_node_t;
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+/****************************************************************************
+ * Name: binary_manager_clear_bin_inode
+ *
+ * Description:
+ *   This function will clear all resource nodes registered in the binary.
+ *
+ * Input parameters:
+ *   bin_idx   -   The index of the binary in binary table
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+void binary_manager_clear_bin_resource(int bin_idx)
+{
+	resource_node_t *resource_node;
+	resource_node_t *node_ptr;
+
+	node_ptr = (resource_node_t *)sq_peek(&BIN_RESOURCELIST(bin_idx));
+	while (node_ptr != NULL) {
+		sq_rem((FAR sq_entry_t *)node_ptr, &BIN_RESOURCELIST(bin_idx));
+		resource_node = node_ptr;
+		if (resource_node->type == BIN_RESOURCE_MQ) {
+			bmvdbg("Unlink MQ %s\n", resource_node->name);
+			mq_unlink(resource_node->name);
+		}
+#ifdef CONFIG_FS_NAMED_SEMAPHORES
+		else if (resource_node->type == BIN_RESOURCE_NAMEDSEM) {
+			bmvdbg("Unlink NAMED SEM %s\n", resource_node->name);
+			sem_unlink(resource_node->name);
+		}
+#endif
+		else {
+			bmdbg("Invalid resource type %d, %s\n", resource_node->type, resource_node->name);
+		}
+		node_ptr = (resource_node_t *)sq_next(node_ptr);
```

Comment:
you can reduce your code to 1 line from 3line
for (node_ptr = sq_peek; node_ptr; node_ptr = sq_next())

Reply:
resource_node needs to be freed after updating node_ptr.

Reply:
I mean your while loop